Fix signal handling in start-l2met-shuttle to ensure clean shutdown.

ojacobson opened the pull request:
Leaving the shell running (04cfd0f) makes the greatest difference. The rest of this change tidies up the process tree.

Commit:
Leave the shell running.

This ensures that the sidecar processes (l2met-shuttle, tee, tee) are properly
terminated when the root of the process group receives SIGTERM or SIGINT by
leaving the shell running to handle those signals. The target process is
instead run as a child of the shell, in parallel with l2met-shuttle, tee, and
tee.

Commit:
Simplify the process tree.

Before (pstree format):

    += bash start-l2met-shuttle
    |--- .heroku/l2met-shuttle/bin/l2met-shuttle http://l2met.example.com/
    \-+- TARGET PROGRAM
      |-+- bash start-l2met-shuttle
      | \--- tee /var/folders/6b/dy5qqkks06dczc7jr7_5jcf4sdqrpy/T/tmp.gHtbbUBr
      \-+- bash start-l2met-shuttle
        \--- tee /var/folders/6b/dy5qqkks06dczc7jr7_5jcf4sdqrpy/T/tmp.gHtbbUBr

After:

    += bash start-l2met-shuttle
    |--- .heroku/l2met-shuttle/bin/l2met-shuttle http://l2met.example.com/
    |--- tee /var/folders/6b/dy5qqkks06dczc7jr7_5jcf4sdqrpy/T/tmp.YaB5eOi6
    |--- tee /var/folders/6b/dy5qqkks06dczc7jr7_5jcf4sdqrpy/T/tmp.YaB5eOi6
    \--- TARGET PROGRAM

tt commented:
I think I made a mistake explaining the problem. It's l2met-shuttle that terminates and leaves the main process up (with no place to log).

@tcrayford, can you confirm?

(I'm not sure why the main process wouldn't die failing to write to the output stream.)

ojacobson commented:
> I think I made a mistake explaining the problem. It's l2met-shuttle that terminates and leaves the main process up (with no place to log).

Aha. That's harder to fix: the tee-to-a-fifo arrangement will never deliver SIGPIPE, which would otherwise be sufficient to cause this whole thing to fall over if l2met-shuttle dies. How would you feel if the process tree became

    += bash start-l2met-shuttle
    |--- .heroku/l2met-shuttle/bin/l2met-shuttle http://l2met.example.com/ # stdout
    |--- .heroku/l2met-shuttle/bin/l2met-shuttle http://l2met.example.com/ # stderr
    \--- TARGET PROGRAM

with each l2met-shuttle handling one or the other of TARGET PROGRAM's standard streams?

Edit: see heroku/l2met-shuttle#6

tt commented:
What do you think of maybe adopting [`start-pgbouncer-stunnel`](https://github.com/heroku/heroku-buildpack-pgbouncer/blob/master/bin/start-pgbouncer-stunnel) or [`start-stunnel`](https://github.com/heroku/heroku-buildpack-redis/blob/master/bin/start-stunnel) wholesale?

I think they fix this problem and they're definitely battle-tested by now.

Commit:
Die on SIGPIPE, too.

Rearranged the process tree so that there are no `tee` processes, which means that if l2met-shuttle exits, then the target process will receive SIGPIPE on the next attempt to write to stdout or stderr. This uses two l2met-shuttle processes (and may lead to odd interleaving, as the two are not synchronous). The process tree now looks like

    |         \-+= TARGET PROCESS
    |           |--- l2met-shuttle --tee URL # stdout
    |           \--- l2met-shuttle --tee URL # stderr

Upgraded l2met-shuttle to 4b4bda3293223100cf3f6d7c34b3545699aa7a3b (built as `GOOS=linux GOARCH=amd64 go build -o l2met-shuttle ./cmd/l2met-shuttle`).

ojacobson commented:
I'll have a look. I just pushed a change that uses `--tee` when launching `l2met-shuttle`, and it _works_ for all the permutations of process death I can think of, but it's worth reusing existing work, here.

ojacobson commented:
Okay. I had a look through the `heroku-buildpack-stunnel` script. I like a lot of things about it, but I think it's solving a different problem, and the use of pipes is more direct for this case.

stunnel, pgbouncer, and friends run as network services, in the background. They'll continue to run indefinitely unless intentionally killed: if a client disconnects, they wait around for a new client to come along. Conversely, if the daemon exits, the caller's on the other side of a network link, and likely has code to try to reconnect, possibly indefinitely, when the right thing to do is to exit.

With `l2met-shuttle`, neither of those situations holds. `l2met-shuttle` reads from stdin and exits if stdin closes for any reason (including the death of the process providing input), so it will exit cleanly if the caller dies. Since it reads from stdin, it can be attached to stdout or stderr directly, causing the upstream process to receive SIGPIPE (which people rarely try to handle, and which causes process death by default) if l2met-shuttle exits.

I'd still be on board with using code from the stunnel buildpack, but it's not obvious to me that the complexity is a win.

kvlou commented:
@ojacobson that is pretty much it, these are more about process management that run on their on, versus consuming stderr/stdout. 

I think it might be more robust to build the wrapper in Go that does the pipe IO.

dmathieu reviewed (approved):
I agree that building the wrapper in Go would be definitely more robust.

Approving anyway, as this fixes a current issue with release phase output.

ojacobson commented:
Forcing changeling, as this repo has no CI configuration.

tt commented:
@ojacobson, you don't need to force. You just need to toggle that you tested it.

tt commented:
I reverted this by force pushing the old commit (as it had already been approved and didn't have to be reviewed again).

tt commented:
```% heroku ps release --app vacuole-staging
=== release (Private-M): curl $LOG_STREAM --silent >/tmp/log-stream && chmod u+x /tmp/log-stream && /tmp/log-stream /bin/sh -c 'start-l2met-shuttle bin/start-stunnel bundle exec bin/dbmigrate' (2)
release.8385: up 2017/05/03 22:28:07 +0200 (~ 10h ago)
release.8808: up 2017/05/03 22:33:10 +0200 (~ 10h ago)

tom-pang commented:
@ojacobson @tt DoD saw a `EPIPE` today, and tracked it down to `l2met-shuttle` having died, but not crashing the dyno. This… doesn't seem to be working?
